### PR TITLE
Df.iat impl

### PR DESCRIPTION
**Pull request by 1e-to**

### CODE REVIEW

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 1]
```diff
@@ -1438,6 +1438,105 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+    Intel Scalable Dataframe Compiler User Guide
+    ********************************************
+
+    Pandas API: pandas.DataFrame.iat
+
+    Examples
+    --------
+    .. literalinclude:: ../../../examples/dataframe_iat.py
+       :language: python
+       :lines: 27-
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: not 28?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 2]
```diff
@@ -1438,6 +1438,105 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+    Pandas DataFrame method :meth:`pandas.DataFrame.iat` implementation.
+
+    .. only:: developer
+        Test: python -m sdc.runtests -k sdc.tests.test_dataframe.TestDataFrame.test_dataframe_iat*
+    """
+
+    _func_name = 'Attribute iat().'
+
+    if not isinstance(self, DataFrameType):
+        raise TypingError('{} The object must be a pandas.dataframe. Given: {}'.format(_func_name, self))
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Why not use `TypeChecker`?

File: sdc/datatypes/hpat_pandas_dataframe_getitem_types.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+
+import pandas
+
+from numba import types, cgutils
+from numba.extending import (models, overload, register_model, make_attribute_wrapper, intrinsic)
+from numba.datamodel import (register_default, StructModel)
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_getitem_types.py]: You don't need brackets here.

File: sdc/hiframes/pd_dataframe_ext.py [hunk 4]
```diff
@@ -52,14 +52,14 @@ class DataFrameAttribute(AttributeTemplate):
...
 
-    def resolve_iloc(self, ary):
-        return DataFrameILocType(ary)
+    # def resolve_iloc(self, ary):
+    #     return DataFrameILocType(ary)
 
-    def resolve_loc(self, ary):
-        return DataFrameLocType(ary)
+    # def resolve_loc(self, ary):
+    #     return DataFrameLocType(ary)
```
Comment (densmirn) [inline sdc/hiframes/pd_dataframe_ext.py]: I vote for removing commented code.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 5]
```diff
@@ -1531,6 +1531,102 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+
+    func_text = '\n'.join(func_lines)
+
+    global_vars = {'pandas': pandas,
+                   'get_dataframe_data': get_dataframe_data}
+
+    return func_text, global_vars
+
+
+def gen_df_getitem_iat_impl(self, row, col, func_name):
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: `gen_df_getitem_iat_impl` -> `gen_df_getitem_tuple_iat_impl`

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 6]
```diff
@@ -1531,6 +1531,102 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+    if not isinstance(self, DataFrameGetitemAccessorType):
+        return None
+
+    accessor = self.accessor.literal_value
+
+    if accessor == 'iat':
+        if isinstance(idx, types.Tuple):
+            row = idx[0].literal_value
+            col = idx[1].literal_value
+            return gen_df_getitem_iat_impl(self.dataframe, row, col, '_df_getitem_tuple_iat_impl')
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Let's hide `'_df_getitem_tuple_iat_impl'` somewhere inside of the implementation generator.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 7]
```diff
@@ -1531,6 +1531,101 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+    Example of generated implementation:
+        def _df_getitem_tuple_iat_impl(self, idx):
+            row, _ = idx
+            data = get_dataframe_data(self._dataframe, 1)
+            res_data = pandas.Series(data)
+            return res_data.iat[row]
+    """
+    func_lines = ['def _df_getitem_tuple_iat_impl(self, idx):',
+                  '  row, _ = idx']
+    if self.columns:
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: What if `if self.columns:` is `False`?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 8]
```diff
@@ -1542,12 +1542,14 @@ def _df_getitem_tuple_iat_impl(self, idx):
...
                   '  row, _ = idx']
-    if self.columns:
+    if col in self.columns:
         func_lines += [
             f"  data = get_dataframe_data(self._dataframe, {col})",
             f"  res_data = pandas.Series(data)",
             f"  return res_data.iat[row]",
         ]
+    else:
+        raise ValueError('Index is out of bounds for axis')
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: I vote for adding unit test to catch such exception.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 9]
```diff
@@ -1529,6 +1529,103 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+
+@sdc_overload(operator.getitem)
+def sdc_pandas_dataframe_accessor_getitem(self, idx):
+    if not isinstance(self, DataFrameGetitemAccessorType):
+        return None
+
+    accessor = self.accessor.literal_value
+
+    if accessor == 'iat':
+        if isinstance(idx, types.Tuple):
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: And if it is not?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 10]
```diff
@@ -1529,6 +1529,103 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+
+    .. only:: developer
+        Test: python -m sdc.runtests -k sdc.tests.test_dataframe.TestDataFrame.test_dataframe_iat*
+    """
+
+    ty_checker = TypeChecker('Attribute iat().')
+    ty_checker.check(self, DataFrameType)
+
+    def sdc_pandas_dataframe_iat_impl(self):
+        return sdc.datatypes.hpat_pandas_dataframe_getitem_types.dataframe_getitem_accessor_init(self, 'iat')
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Could you please import this function and call it without module and etc.?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 11]
```diff
@@ -1529,6 +1529,103 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+def sdc_pandas_dataframe_accessor_getitem(self, idx):
+    if not isinstance(self, DataFrameGetitemAccessorType):
+        return None
+
+    accessor = self.accessor.literal_value
+
+    if accessor == 'iat':
+        if isinstance(idx, types.Tuple):
+            row = idx[0]
+            col = idx[1].literal_value
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Please add check for `idx[1]` to be a literal value

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 12]
```diff
@@ -1529,6 +1529,103 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+            return gen_df_getitem_tuple_iat_impl(self.dataframe, row, col)
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: I believe you don't need codegen here.
```python
def df_getitem_iat_tuple_impl(self, idx):
    row, _ = idx
    data = get_dataframe_data(self._dataframe, col)
    res_data = pandas.Series(data)

    return res_data.iat[row]    
```
should work just fine. And if it doesn't, you could use `literally`:
```python
def df_getitem_iat_tuple_impl(self, idx):
    row, _ = idx
    col_ = literally(col)
    data = get_dataframe_data(self._dataframe, col_)
    res_data = pandas.Series(data)

    return res_data.iat[row]    
```

**Pull request merged**